File: src/js/next/src/Style/ComponentVariant.ts
import type { StyleDefinitionComponentVariant } from '../StyleDefinition.js';
import { Element } from './Element.js';


export class ComponentVariant {
  #data: StyleDefinitionComponentVariant;
  #elements?: readonly Element[];

  constructor(data: StyleDefinitionComponentVariant) {
    this.#data = data;
  }

  elements(): readonly Element[] {
    this.#elements ??= this.#data.elements.map((el) => new Element(el));

    return this.#elements;
  }

  weight(): number {
    return this.#data.weight ?? 1;
  }
}

File: src/go/core/internal/style/component_variant.go
package style

// ComponentVariant is a read-only view over an entry in a component's variants
// block.
type ComponentVariant struct {
	Elements []Element `json:"elements"`
	Weight   *float64  `json:"weight"`
}

// WeightOr1 returns the weighted-pick weight, defaulting to 1.
func (cv ComponentVariant) WeightOr1() float64 {
	if cv.Weight != nil {
		return *cv.Weight
	}
	return 1
}
